### PR TITLE
Allow kernel generator to calculate multiple outputs in single kernel

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Adds functionality to kernel generator to allow calculation of multiple expressions in a single kernel. These expressions need to have results of the same size. 

An option to skip calculation of some results also exists. This will be useful for implementing highly templated functions, such as GLMs.

## Tests

Various combinations of expressions are tested to work in one kernel.

## Side Effects
None.

## Checklist

- [x] Math issue #1342 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/opencl/kernel_generator/multi_result_kernel.hpp [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+  T x;
+  explicit wrapper(T&& x) : x(std::forward<T>(x)) {}
+};
+
+template <typename T>
+wrapper<T> make_wrapper(T&& x) {
+  return wrapper<T>(std::forward<T>(x));
+}
+
+// Pack can only be at the end of the list. We need 2 packs, so we nest structs.
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/multi_result_kernel.hpp]: This comment needs a bit more context.

File: stan/math/opencl/kernel_generator/multi_result_kernel.hpp [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+#include <set>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+
+/**
+ * A wrapper for references. Used to wrap references when putting them in
+ * tuples.
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/multi_result_kernel.hpp]: Not sure a comment needs to reference previous implementations.

File: stan/math/opencl/kernel_generator/multi_result_kernel.hpp [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+    static void add_event(
+        cl::Event e, const std::tuple<wrapper<T_results>...>& results,
+        const std::tuple<wrapper<T_expressions>...>& expressions) {
+      return;
+    }
+  };
+};
+
+template <int n, typename... T_results>
+template <typename... T_expressions>
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/multi_result_kernel.hpp]: Missing comment

File: stan/math/opencl/kernel_generator/multi_result_kernel.hpp [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+      next::add_event(e, results, expressions);
+
+      std::get<n>(expressions).x.add_read_event(e);
+      std::get<n>(results).x.add_write_event(e);
+    }
+  };
+};
+
+template <typename... T_results>
+struct multi_result_kernel_internal<-1, T_results...> {
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/multi_result_kernel.hpp]: Missing comment

File: stan/math/opencl/kernel_generator/calc_if.hpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+template <bool Do_Calculate, typename T,
+          typename = require_all_valid_expressions_and_none_scalar_t<T>>
+inline calc_if_<Do_Calculate, as_operation_cl_t<T>> calc_if(T&& a) {
+  return calc_if_<Do_Calculate, as_operation_cl_t<T>>(
+      as_operation_cl(std::forward<T>(a)));
+}
+
+namespace internal {
+template <typename T>
+struct is_no_output_impl : std::false_type {};
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/calc_if.hpp]: Not sure I like the name is_no_output, but dont have a better suggestion. is_without_output? has_no_output (we tend to use is_* so this is probably bad). is_no_ouput_kernel?

File: stan/math/opencl/kernel_generator/calc_if.hpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  /**
+   * Generates kernel code for assigning this expression into result expression.
+   * @param[in,out] generated set of (pointer to) already generated operations
+   * @param name_gen name generator for this kernel
+   * @param i row index variable name
+   * @param j column index variable name
+   * @param result expression into which result is to be assigned
+   * @return part of kernel with code for this and nested expressions
+   * @throws std::invalid_argument dimensions of expression and result can not
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/opencl/kernel_generator/calc_if.hpp]: Called function `std::get<0>(arguments_).get_whole_kernel_parts(generated, ng, i, j, result)` can throw.

File: stan/math/opencl/kernel_generator/multi_result_kernel.hpp (hunk 3 above)
Comment (t4c1) [reply stan/math/opencl/kernel_generator/multi_result_kernel.hpp]: This is just a definition of a member variable. Why does it need a comment?

Comment (rok-cesnovar) [reply stan/math/opencl/kernel_generator/multi_result_kernel.hpp]: Ah yeah, nevermind.

File: stan/math/opencl/kernel_generator/calc_if.hpp (hunk 5 above)
Comment (t4c1) [reply stan/math/opencl/kernel_generator/calc_if.hpp]: is_without_output makes most sense to me.

**Pull request merged**